Verify that the decoded language is the one we expect before merging

ravisorg opened the pull request:
> This is the result of another evening watching movies and seeing garbage in one of the language fields. The end goal is that Strelingo should serve up perfectly encoded subtitles in the exact language you requested, regardless of the quality of the source subtitles 😁

Verify that the decoded language is the one we expect before merging

Previously we were doing lots of decoding testing in the test suite, and detecting problems and reporting on them. But we were not doing that on the live site when users were requesting translations. As a result, we'd fairly commonly download subtitles and merge them together, even if the decoding function failed to properly decode the correct language. This patch goes about it a little smarter:

- After decoding the subtitles, check that the results contain the language expected.
- Use the [franc-all](https://github.com/wooorm/franc/tree/main) library for language detection, which is far more accurate than just checking the character set.
- If the detected language doesn't match the requested language then we discard the input (it was either corrupted and cannot be decoded, or it was labeled as the wrong language) we return NULL, which bumps us on to the next subtitle file to attempt it instead.

Test suite improvements:
- Added `detectLanguage()` function to identify actual language of content
- Added `skipLanguageValidation` option to decodeSubtitleBuffer() for test analysis
- Updated test to distinguish MISLABELED (wrong language) from FAIL (encoding error)

New test movies added:
- tt0088323 (The NeverEnding Story) - has mislabeled Hungarian subtitle
- tt0116839 (Lawnmower Man 2) - has mislabeled Polish subtitle
- tt0086190 (Star Wars: Return of the Jedi) - original encoding test case
- tt0114558 (Strange Days) - original encoding test case

Test output now shows:
- PASS: Content matches expected language
- MISLABELED: Content decoded OK but is wrong language (with detected language)
- FAIL: Actual encoding failure (replacement chars, etc.)
- KNOWN BAD: Previously identified corrupt files

ravisorg committed:
Since we're being more selective about the inputs we use for the merges now, we should not pre-select 4 results and then merge them. Some (or all) of those 4 might be bad files. Instead we keep trying additional files in that language until we either run out of options, or read our limit of 4.

ravisorg committed:
Removed `zhe` from the list of options
OpenSubtitles uses this for various forms of Chinese + English (ie: dual subtitles). This doesn't make much sense for this service, partially because that can be done with more accuracy by selecting your Chinese variant and English, and partially because it mixes variants together and you never know what you'll get back.
Since and every media ID tested also contained either traditional Chinese (zht) or simplified Chinese (chi) in addition to the zhe version, it makes more sense to offer those two and allow users to mix that with English if desired.
Also made sure it's backward compatible so that if someone had selected the zhe option and has that in their manifest, they'll now get back simplified Chinese instead.

ravisorg committed:
Made error messages clearer when download succeeded but decoding / language validation failed.

ravisorg committed:
Verify that the decoded language is the one we expect before merging

Previously we were doing lots of decoding testing in the test suite, and detecting problems and reporting on them. But we were not doing that on the live site when users were requesting translations. As a result, we'd fairly commonly download subtitles and merge them together, even if the decoding function failed to properly decode the correct language. This patch goes about it a little smarter:

- After decoding the subtitles, check that the results contain the language expected.
- Use the franc-all library for language detection, which is far more accurate than just checking the character set.
- If the detected language doesn't match the requested language then we discard the input (it was either corrupted and cannot be decoded, or it was labeled as the wrong language) we return NULL, which bumps us on to the next subtitle file to attempt it instead.

Test suite improvements:
- Added `detectLanguage()` function to identify actual language of content
- Added `skipLanguageValidation` option to decodeSubtitleBuffer() for test analysis
- Updated test to distinguish MISLABELED (wrong language) from FAIL (encoding error)

New test movies added:
- tt0088323 (The NeverEnding Story) - has mislabeled Hungarian subtitle
- tt0116839 (Lawnmower Man 2) - has mislabeled Polish subtitle
- tt0086190 (Star Wars: Return of the Jedi) - original Thai encoding test case
- tt0114558 (Strange Days) - original Thai encoding test case

Test output now shows:
- PASS: Content matches expected language
- MISLABELED: Content decoded OK but is wrong language (with detected language)
- FAIL: Actual encoding failure (replacement chars, etc.)
- KNOWN BAD: Previously identified corrupt files

ravisorg committed:
Filled in missing test strings for the new test movies introduced in this branch, now that the unit-tests branch is merged in.

ravisorg committed:
Added unit tests for language code aliases (ie: the three romanian codes `ron`, `rum`, and `mol`, as well as the deprecated `zhe` code that maps to `chi` and `zho`) and used the tests to find and fix a bug where `zhe` wasn't converting properly to the aliases.

ravisorg committed:
Fixed related language sets

ravisorg committed:
Strip linebreaks when writing sample content to the server log (ie: don't span multiple lines in the log)

Serkali-sudo commented:
Thats a lof of changes. Looks good though. Merging now. Thank you.

Serkali-sudo commented:
Whoops. My vercel returns 500 error now after merging. Wlil try to fix it.

Serkali-sudo commented:
Okay i fixed it. The problem was i think that franc-all and iso-639-3 are es modules so i guess require() didnt work for them ,so i just loaded them dynamically and 500 error went away.
Hopefully i didnt broke your code when trying to fixing

ravisorg commented:
> Okay i fixed it. The problem was i think that franc-all and iso-639-3 are es modules so i guess require() didnt work for them ,so i just loaded them dynamically and 500 error went away. Hopefully i didnt broke your code when trying to fixing

Totally missed that message, GitHub doesn't seem to notify me 80% of the time 😒 Glad you were able to fix it!

On a related note, I saw you were trying to get cloudflare workers running, and I took a stab at it too because that was something I was thinking of suggesting (probably free, high performance). But it's a lot more complex than expected, mostly because chardet isn't compatible. I switched to jschardet (which is actually much faster anyway) but there are lots of hardcoded chardet references in the libraries. So I put it on the back burner for now.

Serkali-sudo commented:
The reason i tried to swiitch to cloudflare workers was because my vercel free plan is starting to exceed free usage after getting like 20k requests per day. so i was gonna switch to cloudflare workers but i couldnt make it work.

Serkali-sudo commented:
Also because i have a custom domain at cloudflare which also about english learning. Check it out: https://pronouncetube.com/
for example i host this site on cloudflare just fine but i didnt used node js in that site tthoug

ravisorg commented:
I can almost certainly get this app converted to cloudflare workers, but it'll require a bit of a refactor (ie keeping the same functionality but rewriting a lot of the code). If you're ok with that let me know and I'll give it another shot!

ravisorg commented:
> I can almost certainly get this app converted to cloudflare workers, but it'll require a bit of a refactor (ie keeping the same functionality but rewriting a lot of the code). If you're ok with that let me know and I'll give it another shot!

Added an issue for this and assigned myself for now: https://github.com/Serkali-sudo/strelingo-addon/issues/30
